manual/core/metadata/schema/README.md

```diff
@@ -132,7 +132,53 @@ You can also limit the metadata to a subset of keyspaces:
 datastax-java-driver.advanced.metadata.schema.refreshed-keyspaces = [ "users", "products" ]
 ```
 
-If the property is absent or the list is empty, it is interpreted as "all keyspaces".
+Each element in the list can be one of the following:
+
+1. An exact name inclusion, for example `"Ks1"`. If the name is case-sensitive, it must appear in
+   its exact case.
+2. An exact name exclusion, for example `"!Ks1"`.
+3. A regex inclusion, enclosed in slashes, for example `"/^Ks.*/"`. The part between the slashes
+   must follow the syntax rules of [java.util.regex.Pattern].
+4. A regex exclusion, for example `"!/^Ks.*/"`.
+
+If the list is empty, or the option is unset, all keyspaces will match. Otherwise:
+
+* If a keyspace matches an exact name inclusion, it is always included, regardless of what any other
+  rule says.
+* Otherwise, if it matches an exact name exclusion, it is always excluded, regardless of what any
+  regex rule says.
+* Otherwise, if there are regex rules:
+
+  * if they're only inclusions, the keyspace must match at least one of them.
+  * if they're only exclusions, the keyspace must match none of them.
+  * if they're both, the keyspace must match at least one inclusion and none of the
+    exclusions.
+
+For example, given the keyspaces `system`, `ks1`, `ks2`, `data1` and `data2`, here's the outcome of
```

Comment:
The inclusion/exclusion rules are very powerful, yet slightly convoluted. Do we absolutely need to distinguish "exact inclusions" from "regex inclusions"? After all, an exact keyspace name is just a special case of regex that has no wildcards. If we could consider that all the elements are regexes, we could probably simplify the rules and also avoid the usage of the special syntax `/.../` for regexes.

Reply:
From a semantic point of view, exact matches have a different meaning:

* `[ "^/system_.*/", "system.traces" ]`: I think most people would expect this to mean "exclude all system keyspaces, but include `system_traces`.
* on the other hand, if we have two regexes: `[ "/ks.*/", "^/.*2/" ]`. Then it's fair to assume an intersection: all keyspaces that start with `ks`, but do not end with `2`.

If I apply the "two regexes" rule to the first example, I get "all keyspaces named `system_traces`, but that do not start with `system_`". That's the empty set.

Reply:
I'm still not 100% convinced. If I wanted to "exclude all system keyspaces, but include `system_traces`" I could simply specify `[ "system_traces" ]` right? For your first example `[ "!/system_.*/", "system.traces" ]`, returning an empty set seem legit to me.

Reply:
One thing that is maybe not obvious is that `[ "!/system_.*/", "system.traces" ]` also implies "include everything that doesn't start with system_" (per the third regex rule: if they're only exclusions, the keyspace must match none of them).

If you simply specify `[ "system_traces" ]` you only get system_traces. And this one can't really change because we need to preserve backward compatibility with the 4.9 behavior.

Returning the empty set for `[ "!/system_.*/", "system.traces" ]` seems very wrong to me. If the user mentioned a specific keyspace by name, it's almost certainly because they wanted to include it. That's why exact names win over any other rule.

I think it will take too much time to discuss this on individual examples: if you have a better idea, please suggest a complete set of rules and I'll try to poke holes in it 😉 
I already went over several versions on my own, it will probably be one of them.

Or maybe patterns are overkill. We could simply hard-code the system exclusions when nothing is specified, and keep the current behavior (exact inclusions only) otherwise.

Reply:
Since we're running out of time I'm leaving the rules as they are for now. This is not strictly part of the API so I guess we could revisit the rules later on if necessary.